src/zarr/buffer/core.py

```diff
@@ -179,10 +182,16 @@ def from_buffer(cls, buffer: Buffer) -> Self:
         Returns
         -------
             A new buffer representing the content of the input buffer
+
+        Note
+        ----
+        Subclasses of `Buffer` must override this method to implement
+        more optimal conversions that avoid copies where possible
         """
-        return cls.from_array_like(buffer.as_array_like())
+        ...
 
     @classmethod
+    @abstractmethod
```

Comment:
test failures are due to this method becoming abstract, while it still gets called in `to_buffer_dict` in `metadata.py`. Two things should change to fix this: first, I think maybe the order of the decorators here should be flipped to ensure that `Buffer.from_bytes` can't be called without an exception, and second `metadata.py` needs to not be calling any `Buffer` methods.

Reply:
I'm not sure we can flip the order of decorators here based on this snippet from the `abstractmethod` [docs](https://docs.python.org/3/library/abc.html#abc.abstractmethod):
> When abstractmethod() is applied in combination with other method descriptors, it should be applied as the innermost decorator

I agree though that `metadata.py` shouldn't be calling any `Buffer` methods and should instead be calling `prototype.buffer` methods or `default_buffer_prototype.buffer` methods. Is there a preference between propagating a `prototype` argument up the call stack or just using `default_buffer_prototype` since these `to_bytes` calls are not in the critical performance path?

Reply:
Resolved the CI issues by tracking all calls to abstract classmethods of `Buffer` and `NDBuffer`. Here are the main ones and the current solution:
- `metadata.py`: switch to using `default_buffer_prototype.buffer`
- `group.py`: switch to using `default_buffer_prototype.buffer`
- `sharding.py`: switch to using `default_buffer_prototype.buffer`
- `codecs/_v2.py`: switch to using `cpu.Buffer` and `cpu.NDBuffer` since these are all explicitly CPU only.